pkg/streamer/reader_test.go

```diff
@@ -0,0 +1,689 @@
+	err := os.MkdirAll(relayDir, 0744)
+	c.Assert(err, IsNil)
+	f, err := os.OpenFile(fullPath, os.O_CREATE|os.O_WRONLY, 0644)
+	c.Assert(err, IsNil)
+	defer f.Close()
+	_, err = f.Write(replication.BinLogFileHeader)
+	c.Assert(err, IsNil)
+	for _, ev := range baseEvents {
+		_, err = f.Write(ev.RawData)
+		c.Assert(err, IsNil)
+	}
+
+	// invalid UUID in UUID list, error
+	r.uuids = []string{currentUUID, "invalid.uuid"}
+	ctx1, cancel1 := context.WithTimeout(context.Background(), time.Second)
+	defer cancel1()
+	needSwitch, needReParse, latestPos, nextUUID, nextBinlogName, err := r.parseFile(
+		ctx1, s, filename, offset, relayDir, firstParse, currentUUID, possibleLast)
+	c.Assert(err, ErrorMatches, ".*not valid.*")
+	c.Assert(needSwitch, IsFalse)
+	c.Assert(needReParse, IsFalse)
+	c.Assert(latestPos, Equals, int64(0))
+	c.Assert(nextUUID, Equals, "")
+	c.Assert(nextBinlogName, Equals, "")
+	t.purgeStreamer(c, s)
+
+	// next sub dir exits, need to switch
+	r.uuids = []string{currentUUID, switchedUUID}
+	err = os.MkdirAll(nextRelayDir, 0744)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(nextFullPath, replication.BinLogFileHeader, 0644)
+	c.Assert(err, IsNil)
+
+	// has relay log file in next sub directory, need to switch
+	ctx2, cancel2 := context.WithTimeout(context.Background(), time.Second)
+	defer cancel2()
+	needSwitch, needReParse, latestPos, nextUUID, nextBinlogName, err = r.parseFile(
+		ctx2, s, filename, offset, relayDir, firstParse, currentUUID, possibleLast)
+	c.Assert(err, IsNil)
+	c.Assert(needSwitch, IsTrue)
+	c.Assert(needReParse, IsFalse)
+	c.Assert(latestPos, Equals, int64(0))
+	c.Assert(nextUUID, Equals, switchedUUID)
+	c.Assert(nextBinlogName, Equals, nextFilename)
+	t.purgeStreamer(c, s)
+
+	// NOTE: if we want to test the returned `needReParse` of `needSwitchSubDir`,
+	// then we need to mock `fileSizeUpdated` or inject some delay or delay.
+}
+
+func (t *testReaderSuite) TestParseFileRelayWithIgnorableError(c *C) {
+	var (
+		filename     = "test-mysql-bin.000001"
+		baseDir      = c.MkDir()
+		offset       int64
+		firstParse   = true
+		possibleLast = true
+		baseEvents   = t.genBinlogEvents(c, 0)
+		currentUUID  = "b60868af-5a6f-11e9-9ea3-0242ac160006.000001"
+		relayDir     = filepath.Join(baseDir, currentUUID)
+		fullPath     = filepath.Join(relayDir, filename)
+		s            = newLocalStreamer()
+		cfg          = &BinlogReaderConfig{RelayDir: baseDir}
+		r            = NewBinlogReader(cfg)
+	)
+
+	// create the current relay log file and write some events
+	err := os.MkdirAll(relayDir, 0744)
+	c.Assert(err, IsNil)
+	f, err := os.OpenFile(fullPath, os.O_CREATE|os.O_WRONLY, 0644)
+	c.Assert(err, IsNil)
+	defer f.Close()
+	_, err = f.Write(replication.BinLogFileHeader)
+	c.Assert(err, IsNil)
+	for _, ev := range baseEvents {
+		_, err = f.Write(ev.RawData)
+		c.Assert(err, IsNil)
+	}
+	_, err = f.Write([]byte("some invalid binlog event data"))
+	c.Assert(err, IsNil)
+
+	// EOF error ignored
```

Comment:
can we add one more test case to test `parseFile` with the ignorable error `io.EOF`

Reply:
added in https://github.com/pingcap/dm/pull/140/commits/ec91143d6bb3dea85021f1f3384bff2a51a02148